### PR TITLE
feat(@netlify/vite-plugin-react-router): add edge support

### PR DESCRIPTION
**Pull request by serhalp**
## Summary

This PR adds opt-in support for deploying React Router 7 apps to Netlify Edge Functions (Deno runtime) via the `edge: boolean` plugin option. Previously, the plugin only supported Netlify Functions (Node.js runtime).

Toward FRB-1519

## Changes

- Add `edge?: boolean` option to Vite plugin factory
- When `edge` is enabled:
  - Configure Vite SSR build with `target: 'webworker'` and other Deno-compatible build config
  - Generate an edge function in `.netlify/v1/edge-functions/`, instead of a function `.netlify/v1/functions/`
- Add `excludedPaths?: string[]` option to allow users to exclude paths from the React Router handler. Required when using `edge: true` and the project also has its own Netlify Functions with a configured `path`, otherwise only the RR handler runs on all dynamic paths.

## Testing

- Unskip the RR7 edge e2e suite
- Add a few tests I noticed were missing for edge (or vice versa)
- I thought I'd add tests to ensure prerendering works while I was here (this was added in RR7 but did not exist in Remix 2)... but unfortunately I discovered that this does not in fact work. I left the tests and marked them as failing.
- You can also link this locally with https://github.com/netlify/react-router-template/ and confirm it works ✅

## Documentation

I added complete documentation for opting in (and back out of) edge deployment.

## Implementation

Edge support is more involved than regular serverless functions for three main reasons:

1. EFs don't have a `preferStatic` option like NFs.
    - Instead, to avoid needlessly running compute to serve static assets, we must populate the EF's `excludedPath` with all published client assets. This skips the EF entirely for these paths.
2. Although there is a precise processing order for EFs and for NFs, [all EFs run before any NF](https://docs.netlify.com/start/core-concepts/request-chain/).
    - This means that since our generated EF runs on `path: '/*'`, if a user has their own NF with `path: '/foo'`, our EF will run first and render a response from React Router and the user's NF will never run.
    - To that end, this PR adds and documents an `excludedPaths?: []` option to the plugin.
    - There were various alternative solutions here, but this seemed like a reasonable approach for now.
3. React Router 7 [contains only a single default built-in server entry](https://github.com/remix-run/react-router/tree/89e2bfe8232ad7c24cf610b0998e9028dcfa9ce8/packages/react-router-dev/config/defaults) and this assumes a Node.js environment.
    - Unfortunately, this means when using `edge: true` the user must include their own `app/entry.server.tsx` that uses `renderToReadableStream` instead of `renderToPipeableStream`.
    - There may be some magic we can implement here to make this work out of the box. It didn't seem worth it to me to fight against the framework here. We can also always add this later.

### CODE REVIEW

File: packages/vite-plugin-react-router/package.json [hunk 1]
```diff
@@ -16,6 +16,14 @@
         "types": "./dist/index.d.mts",
         "default": "./dist/index.mjs"
       }
+    },
+    "./function-handler": {
+      "types": "./dist/function-handler.d.mts",
+      "default": "./dist/function-handler.mjs"
+    },
+    "./edge-function-handler": {
```
Comment (serhalp) [inline packages/vite-plugin-react-router/package.json]: Check out the first lines of `FUNCTION_HANDLER` and `EDGE_FUNCTION_HANDLER` to understand what these are for. They're the _runtime_ entry points.

(This should probably always have been an explicit separate export, honestly.)

File: packages/vite-plugin-react-router/README.md [hunk 2]
```diff
@@ -38,6 +37,115 @@ export default defineConfig({
...
+Second, you **must** provide an `app/entry.server.tsx` (or `.jsx`) file that uses web-standard APIs compatible with the
+Deno runtime. Create a file with the following content:
+
+> [!IMPORTANT]
+>
+> This file uses `renderToReadableStream` (Web Streams API) instead of `renderToPipeableStream` (Node.js API), which is
+> required for the Deno runtime. You may customize your server entry file, but see below for important edge runtime
+> constraints.
+
+```tsx
```
Comment (serhalp) [inline packages/vite-plugin-react-router/README.md]: identical to the Cloudflare one in the RR7 codebase... https://github.com/remix-run/react-router/blob/cb9a090316003988ff367bb2f2d1ef5bd03bd3af/integration/helpers/vite-plugin-cloudflare-template/app/entry.server.tsx

File: packages/vite-plugin-react-router/src/plugin.ts [hunk 3]
```diff
@@ -75,24 +138,49 @@ export function netlifyPlugin(): Plugin {
...
+        if (edge) {
+          // Edge Functions do not have a `preferStatic` option, so we must exhaustively exclude
+          // static files to serve them from the CDN without compute.
+          // RR7's build out dir contains /server and /client subdirectories. This is documented and
+          // not configurable, so the client out dir is always at ../client from the server out dir.
+          const clientDir = join(resolvedConfig.build.outDir, '..', 'client')
+          const entries = await readdir(clientDir, { withFileTypes: true })
+          const excludedPath = [
+            '/.netlify/*',
+            ...entries.map((entry) => (entry.isDirectory() ? `/${entry.name}/*` : `/${entry.name}`)),
```
Comment (pieh) [inline packages/vite-plugin-react-router/src/plugin.ts]: If we only check top-level files and dirs in `build/client`, then at least in my case it would exclude `/assets/*` which means users won't be allowed to have their own routes that start with `/assets/` segment - given this is in new code path for edge rendering, this at least won't cause regressions, but it might lead to very confusing problems for users opting to use edge rendering

Comment (pieh) [reply packages/vite-plugin-react-router/src/plugin.ts]: > ... ah but wait! You're saying actually there's nothing special about the dir assets/. A project could have foo/bar.jpg and we'd exclude all of foo/* from the function. hmm.

I don't know enough about react-router - I'm not sure if it's possible to inject "random" files to be available (kind of like [`public` directory in Next.js](https://nextjs.org/docs/pages/api-reference/file-conventions/public-folder)) or wether it would only ever be `assets` directory in practice?

If it would only ever be `assets`, then I think it's reasonable enough to accept tradeoff here about not overloading `assets` path segment by users / playing with fire and to not explode number of excluded paths in configuration that will happen when we need to list all the individual files, but I will leave decision to you here. Just wanted to flag this as potential problem

File: packages/vite-plugin-react-router/src/index.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-export type { GetLoadContextFunction, RequestHandler } from './server'
-export { createRequestHandler, netlifyRouterContext } from './server'
+export type { GetLoadContextFunction, RequestHandler } from './function-handler'
+export { createRequestHandler, netlifyRouterContext } from './function-handler'
```
[the remark this answers is not in the document]
Comment (serhalp) [reply packages/vite-plugin-react-router/src/index.ts]: Awesome, thank you for catching that! I've pushed a fix and reworked how I was segregating edge vs. serverless. It feels better to me now!

File: packages/vite-plugin-react-router/src/plugin.ts (hunk 3 above)
Comment (serhalp) [reply packages/vite-plugin-react-router/src/plugin.ts]: Ah, yep, good follow-up. It does look like `assets` is hardcoded in RR. So I think I'll leave this as is.

File: packages/vite-plugin-react-router/src/plugin.ts [hunk 5]
```diff
@@ -29,20 +60,39 @@ export default createRequestHandler({
...
 // This is written to the functions directory. It just re-exports
 // the compiled entrypoint, along with Netlify function config.
-function generateNetlifyFunction(handlerPath: string) {
+function generateNetlifyFunction(handlerPath: string, excludedPath: Array<string>) {
   return /* js */ `
     export { default } from "${handlerPath}";
 
     export const config = {
-      name: "React Router server handler",
+      displayName: "React Router server handler",
```
[the remark this answers is not in the document]
Comment (serhalp) [reply packages/vite-plugin-react-router/src/plugin.ts]: Oof, you're right. I'll undo this and fix the docs.
